libyul/backends/evm/OptimizedEVMCodeTransform.cpp

```diff
@@ -0,0 +1,527 @@
+		}
+	);
+	yulAssert(m_assembly.stackHeight() == static_cast<int>(m_stack.size()), "");
+}
+
+void OptimizedEVMCodeTransform::operator()(CFG::BasicBlock const& _block)
+{
+	// Assert that this is the first visit of the block and mark as generated.
+	yulAssert(m_generated.insert(&_block).second, "");
+
+	auto const& blockInfo = m_stackLayout.blockInfos.at(&_block);
+
+	// Assert that the stack is valid for entering the block.
+	assertLayoutCompatibility(m_stack, blockInfo.entryLayout);
+	m_stack = blockInfo.entryLayout; // Might set some slots to junk, if not required by the block.
+	yulAssert(static_cast<int>(m_stack.size()) == m_assembly.stackHeight(), "");
+
+	// Emit jump label, if required.
+	if (auto label = util::valueOrNullptr(m_blockLabels, &_block))
+		m_assembly.appendLabel(*label);
+
+	for (auto const& operation: _block.operations)
+	{
+		// Create required layout for entering the operation.
+		createStackLayout(m_stackLayout.operationEntryLayout.at(&operation));
+
+		// Assert that we have the inputs of the operation on stack top.
+		yulAssert(static_cast<int>(m_stack.size()) == m_assembly.stackHeight(), "");
+		yulAssert(m_stack.size() >= operation.input.size(), "");
+		size_t baseHeight = m_stack.size() - operation.input.size();
+		assertLayoutCompatibility(
+			m_stack | ranges::views::take_last(operation.input.size()) | ranges::to<Stack>,
+			operation.input
+		);
+
+		// Perform the operation.
+		std::visit(*this, operation.operation);
+
+		// Assert that the operation produced its proclaimed output.
+		yulAssert(static_cast<int>(m_stack.size()) == m_assembly.stackHeight(), "");
+		yulAssert(m_stack.size() == baseHeight + operation.output.size(), "");
+		yulAssert(m_stack.size() >= operation.output.size(), "");
+		assertLayoutCompatibility(
+			m_stack | ranges::views::take_last(operation.output.size()) | ranges::to<Stack>,
+			operation.output
+		);
+	}
+
+	// Exit the block.
+	std::visit(util::GenericVisitor{
+		[&](CFG::BasicBlock::MainExit const&)
+		{
+			m_assembly.appendInstruction(evmasm::Instruction::STOP);
+		},
+		[&](CFG::BasicBlock::Jump const& _jump)
+		{
+			// Create the stack expected at the jump target.
+			createStackLayout(m_stackLayout.blockInfos.at(_jump.target).entryLayout);
+
+			// If this is the only jump to the block, we do not need a label and can directly continue with the target block.
+			if (!m_blockLabels.count(_jump.target) && _jump.target->entries.size() == 1)
+			{
+				yulAssert(!_jump.backwards, "");
+				(*this)(*_jump.target);
+			}
+			else
+			{
+				// Generate a jump label for the taret, if not already present.
+				if (!m_blockLabels.count(_jump.target))
+					m_blockLabels[_jump.target] = m_assembly.newLabelId();
+
+				// If we already have generated the target block, jump to it, otherwise generate it in place.
+				if (m_generated.count(_jump.target))
+					m_assembly.appendJumpTo(m_blockLabels[_jump.target]);
+				else
+					(*this)(*_jump.target);
```

Comment:
This creates one C++ stack frame per jump, right? Could this lead to stack exhaustion?

Reply:
Theoretically, yes, but I don't think you can get there in practice.

Reply:
We had this problem in the old cpp-ethereum evm interpreter a lot, but maybe the stack frames then were also just very large.

Reply:
I can always rewrite it to something non-recursive, but my guess would be that it'd become much harder to read...

Reply:
But yeah, the stack frame at the recursions here is largely empty, not sure it's an issue.